src/odemis/gui/comp/overlay/world.py

```diff
@@ -1231,46 +1314,350 @@ def draw(self, ctx, selected, shift=(0, 0), scale=1.0, canvas=None):
+        dx, dy = self.p_end_pos[0] - self.p_start_pos[0], self.p_end_pos[1] - self.p_start_pos[1]
+
+        # unit vector for buffer (pixel) coordinates
+        dpx, dpy = b_end[0] - b_start[0], b_end[1] - b_start[1]
+
+        phi = math.atan2(dx, dy) % (2 * math.pi)  # phi angle  of the label line in radians
+
+        # Find the label line length by calculating the Euclidean distance
+        pixel_length = math.hypot(dpx, dpy)  # label lime length in pixels
+
+        # text is placed horizontal
+        self._label.deg = math.degrees(math.pi)
+
+        # Draws a black background for the ruler
+        ctx.set_line_width(2)
+        ctx.set_source_rgba(0, 0, 0, 0.5)
+        ctx.move_to(*b_start)
+        ctx.line_to(*b_end)
+
+        # Ruler of 1 pixel width. Highlight the selected ruler and make it slightly thicker (2 pixels)
+        if selected:
+            ctx.set_source_rgba(*self.highlight)
+            ctx.set_line_width(2)
+        else:
+            ctx.set_source_rgba(*self.colour)
+            ctx.set_line_width(1)
+        ctx.set_line_join(cairo.LINE_JOIN_MITER)
+
+        self._label.weight = cairo.FONT_WEIGHT_NORMAL
+        self._label.colour = self.highlight if selected else self.colour
+
+        self._label.text = self.text
+
+        # Display text at the end of the line
+        # Cairo renders text from the bottom left, but we want to treat different points as the origin.
+        # φ ~ 0 --> move the label to the left for width/2 & up for height
+        # φ ~ 180 --> move the label to the left for width/2
+        # 0 < φ < 180 --> move the label down for height/2
+        # 180 < φ < 360 --> move the label to the left for width & down for height/2
+
+        self._label.pos = Vec(b_end[0], b_end[1])
+
+        if phi < math.pi / 4:
+            self._label.align = wx.ALIGN_BOTTOM | wx.ALIGN_CENTER_HORIZONTAL
+        elif phi > 7 * math.pi / 4:
+            self._label.align = wx.ALIGN_BOTTOM | wx.ALIGN_CENTER_HORIZONTAL
+        elif 3 * math.pi / 4 < phi < 5 * math.pi / 4:
+            self._label.align = wx.ALIGN_CENTER_HORIZONTAL
+        elif math.pi / 4 < phi < 3 * math.pi / 4:
+            self._label.align = wx.ALIGN_CENTRE_VERTICAL
+        else:  # math.pi < phi < 2 * math.pi
+            self._label.align = wx.ALIGN_RIGHT | wx.ALIGN_CENTER_VERTICAL
+
+        self._label.font_size = 14
+        if font_size:
+            # When exporting a spatial view for print-ready the font size to draw on the fake canvas is given
+            self._label.font_size = font_size
+
+        # If the label is smaller than 1 pixel, make it seem as 1 point (1 pixel)
+        if pixel_length <= 1:
+            ctx.move_to(*b_start)
+            ctx.line_to(b_start[0] + 1, b_start[1] + 1)
+        else:
+            ctx.move_to(*b_start)
+            ctx.line_to(*b_end)
+        ctx.stroke()
+
+        self._label.draw(ctx)
+        self._update_view_positions(b_start, b_end)
```

Comment:
You shouldn't call `_update_view_positions()` all the time. Maybe just move the `if` from the top to here? Or does the should the update happen before the drawing? I have a feeling this is only used for the mouse interaction, but I might be wrong.
An alternative could be to put the burden on the caller. You could have a 6th public method `sync_with_canvas()` (or `on_canvas_moved()`), which would do what `_update_view_positions()` does. GagdetOverlay would be in charge of calling it (and for now, you could just do it after every draw). One of the main advantage would be that you could go away with this weird mix-up with `canvas` argument.